### PR TITLE
Task/bundle generated extensions

### PR DESCRIPTION
**Pull request by jobala**
## Purpose

This PR generates cli-extensions and bundles them in the msi package during build.

## How It Works
This PR introduces changes that allow us to generate, build and install extensions. 

The `generate_extensions.py` script generates extensions from **Open API** descriptions in `open-api/beta`. It first generates configuration files that **autorest.az** and **autorest.python** uses to generate extensions.

The next step is to build the generated extensions, which is done by the `build_exstensions.py` script. It builds the extensions into distributable python packages. The script also creates a manifest of installed extensions which the CLI uses for loading command modules.

Finally, extensions are installed by the `install_extensions.py`  script. We also run this script when building `msgraph-cli`. 

## Known Issues

We are not generating extensions with long names,  tracked in #AB 5222
Some extensions fail to generate, tracked in #AB 5223
Error stacktrace shown to the user, tracked in #AB 5224

### CODE REVIEW

File: build_extensions.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""
+This script builds generated extensions ands adds built extensions to a manifest file.
```
Comment (ddyett) [inline build_extensions.py]:
```suggestion
This script builds generated extensions and adds built extensions to a manifest file.
```

File: build_extensions.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import os
+import subprocess
+import sys
+from os import path
+from setuptools import sandbox
+
+path_to_extensions = path.join(os.getcwd(), 'msgraph-cli-extensions', 'src')
+path_to_manifest = path.join(os.getcwd(), 'src',
+                             'msgraph-cli-core', 'msgraph', 'cli', 'core', 'installed_extensions.py')
```
Comment (ddyett) [inline build_extensions.py:14]: do we need the msgraph folder?

File: build_scripts/windows/scripts/build.cmd [hunk 3]
```diff
@@ -10,10 +10,10 @@ if "%CLI_VERSION%"=="" (
   goto ERROR
 )
 
-set PYTHON_VERSION=3.8.3
+set PYTHON_VERSION=3.6.6
```
Comment (ddyett) [inline build_scripts/windows/scripts/build.cmd:13]: why the change to a lower version?

File: generate_extensions.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def generate_extension_from_open_api_description():
+    open_api_descriptions = get_open_api_descriptions()
+
+    for item in open_api_descriptions:
+        file_name, file_path = item
+        file_name = remove_file_extension(file_name)
+
+        if len(file_name) < 15:
```
Comment (ddyett) [inline generate_extensions.py:15]: what's the significance of 15.  maybe use a constant name to help clarify meaning.

File: generate_extensions.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            # Config files are used to modify generated extensions
+            generate_az_config_for(file_name)
+            generate_cli_config_for(file_name)
+            generate_python_config_for(file_name)
+
+            subprocess.run(
+                ['autorest-beta', '--az', '--v3', f'''--input-file:{file_path}''',
+                 r'''--azure-cli-extension-folder=msgraph-cli-extensions''',
+                 r'''--use=@autorest/python@5.1.0-preview.2''',
```
Comment (ddyett) [inline generate_extensions.py:24]: lots of specific version numbers we'll have to keep updating.

File: generate_extensions.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                 r'''--use=@autorest/modelerfour@4.14.366''',
+                 r'''--use=https://github.com/Azure/autorest.az/releases/download/1.4.1-b.20200623.1/autorest-az-1.4.1.tgz'''],
+                shell=True)
+
+
+def get_open_api_descriptions():
+    result = []
+
+    open_api_dir = path.join(os.getcwd(), 'open-api-docs', 'beta')
```
Comment (ddyett) [inline generate_extensions.py:33]: do we have an issue tracking V1 support?

File: generate_extensions.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    open_api_files = os.listdir(open_api_dir)
+
+    for file in open_api_files:
+        file_and_path = (file, path.join(open_api_dir, file))
+        result.append(file_and_path)
+
+    return result
+
+
+def remove_file_extension(file_name):
```
Comment (ddyett) [inline generate_extensions.py]: this isn't just removing the file extension it's in the groups case also removing that so function can be better named.  Are you removing these for naming limits?

File: src/msgraph-cli-core/msgraph/cli/core/__init__.py [hunk 8]
```diff
@@ -91,6 +92,10 @@ def load_arguments(self, command=None):
                 loader._update_command_definitions()  # pylint: disable=protected-access
 
     def _update_command_table_from_modules(self, args):
+        """Loads command_table from msgraph.cli.command_modules
```
Comment (ddyett) [inline src/msgraph-cli-core/msgraph/cli/core/__init__.py:95]: is it right to think about these command modules as custom modules we write?

File: src/msgraph-cli-core/msgraph/cli/core/__init__.py [hunk 9]
```diff
@@ -126,11 +149,19 @@ def __init__(self, cli_ctx=None, command_group_cls=None, argument_context_cls=No
...
+
+        :param group_name:   group_name of the set of commands. ie users
+        :param command_type: Cli command_type
+        '''
         if command_type:
             kwargs['command_type'] = command_type
         return self._command_group_cls(self, group_name, **kwargs)
 
     def _cli_command(self, name, operation=None, handler=None, argument_loader=None, description_loader=None, **kwargs):
+        '''Adds a command to the command table
```
Comment (ddyett) [inline src/msgraph-cli-core/msgraph/cli/core/__init__.py:162]: appreciate the documentation

File: src/msgraph-cli-core/msgraph/cli/core/installed_extensions.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
+installed_extensions = [ 'azext_accessreview',
```
Comment (ddyett) [inline src/msgraph-cli-core/msgraph/cli/core/installed_extensions.py:1]: why do we need these?

File: build_scripts/windows/scripts/build.cmd (hunk 3 above)
Comment (jobala) [reply build_scripts/windows/scripts/build.cmd:13]: I changed to a lower version so that the build works. The Azure hosted Python version is 3.6.6. We are using the hosted Python  version in the build process. There's a work item to upgrade to the latest Python version.

File: generate_extensions.py (hunk 6 above)
Comment (jobala) [reply generate_extensions.py:33]: There are no open-api descriptions under the v1 folder.

File: generate_extensions.py (hunk 7 above)
Comment (jobala) [reply generate_extensions.py]: Yes, I am removing these for naming limits.

File: src/msgraph-cli-core/msgraph/cli/core/installed_extensions.py (hunk 10 above)
Comment (jobala) [reply src/msgraph-cli-core/msgraph/cli/core/installed_extensions.py:1]: I considered two approaches.

1. Load the extensions directly from msgraph-cli-extensions.
2. Create a manifest of installed extensions.

The challenge I faced with the first approach is that folder structure for generated extensions was not suitable for loading directly from path.

I used the second approach because it offered convenience when loading extensions and is explicit about what extensions are loaded.

NOTE: This file is autogeneratad when running `build_extensions.py`

File: build_extensions.py (hunk 2 above)
Comment (jobala) [reply build_extensions.py:14]: Yes, `msgraph` folder is our namespace package.

File: src/msgraph-cli-core/msgraph/cli/core/__init__.py (hunk 8 above)
Comment (jobala) [reply src/msgraph-cli-core/msgraph/cli/core/__init__.py:95]: I preferred to keep a logical division between command modules we write and those that are generated because it gives us the affordance to update generated command modules without having to update the cli.
